File: CVersion/Ship.cpp
#include <string>
#include <vector>
#include "Ship.h"

using namespace std;

Ship::Ship(int r, int c, Orientation o, int len) : r(r), c(c), o(o), len(len), rows(compute_rows()), cols(compute_cols()) {};

bool Ship::operator==(const Ship &other) const
{
    return r == other.r && c == other.c && o == other.o && len == other.len;
}

vector<int> Ship::compute_rows()
{
    vector<int> rows(Ship::len);
    if (Ship::o == Orientation::H)
    {
        for (int i = 0; i < Ship::len; ++i)
        {
            rows[i] = Ship::r;
        }
    }
    else
    {
        for (int i = 0; i < Ship::len; ++i)
        {
            rows[i] = Ship::r + i;
        }
    }
    return rows;
}

vector<int> Ship::compute_cols()
{
    vector<int> cols(Ship::len);
    if (Ship::o == Orientation::H)
    {
        for (int i = 0; i < Ship::len; ++i)
        {
            cols[i] = Ship::c + i;
        }
    }
    else
    {
        for (int i = 0; i < Ship::len; ++i)
        {
            cols[i] = Ship::c;
        }
    }
    return cols;
}

File: PythonVersion/Ship.py
import numpy as np
from enum import Enum 
from collections import namedtuple

class Orientation(Enum):
        H = 1
        V = 2

Ship = namedtuple("Ship", ["r", "c", "o", "length", "rows", "cols"])

def rows(o,r, length):
    if o is Orientation.H:
        # horizontal: row fixed = r, cols vary
        return tuple(np.full(length, r))
    else:
        # vertical: rows vary starting at r
        return tuple(np.arange(r, r + length))

def cols(o,c, length):
    if o is Orientation.V:
        # horizontal: row fixed = r, cols vary
        return tuple(np.full(length, c))
    else:
        # vertical: rows vary starting at r
        return tuple(np.arange(c, c + length))

